src/tree/split_evaluator.cc

```diff
@@ -302,5 +305,194 @@ XGBOOST_REGISTER_SPLIT_EVALUATOR(MonotonicConstraint, "monotonic")
+
+/*! \brief Enforces that the tree is monotonically increasing/decreasing with respect to a user specified set of
+      features.
+*/
+class InteractionConstraint final : public SplitEvaluator {
+ public:
+  explicit InteractionConstraint(std::unique_ptr<SplitEvaluator> inner) {
+    if (!inner) {
+      LOG(FATAL) << "InteractionConstraint must be given an inner evaluator";
+    }
+    inner_ = std::move(inner);
+  }
+
+  void Init(const std::vector<std::pair<std::string, std::string> >& args)
+      override {
+    inner_->Init(args);
+    params_.InitAllowUnknown(args);
+    Reset();
+  }
+
+  void Reset() override {
+    // Parse interaction constraints
+    {
+      if (params_.interaction_constraints == "NULL") {
+        // no constraint supplied: any feature can interact with any feature
+        interaction_constraints_.emplace_back();
+        interaction_constraints_[0].reserve(params_.num_feature);
+        for (bst_uint i = 0; i < params_.num_feature; ++i) {
+          interaction_constraints_[0].insert(i);
+        }
+      } else {
+        std::istringstream iss(params_.interaction_constraints);
+        dmlc::JSONReader reader(&iss);
+        // Read std::vector<std::vector<bst_uint>> first and then
+        //   convert to std::vector<std::unordered_set<bst_uint>>
+        std::vector<std::vector<bst_uint>> tmp;
+        reader.Read(&tmp);
+        for (const auto& e : tmp) {
+          interaction_constraints_.emplace_back(e.begin(), e.end());
+        }
+      }
+    }
+
+    // Initialise interaction constraints record with all variables permitted for the first node
+    int_cont_.clear();
+    int_cont_.resize(1, std::unordered_set<bst_uint>());
+    int_cont_[0].reserve(params_.num_feature);
+    for (bst_uint i = 0; i < params_.num_feature; ++i) {
+      int_cont_[0].insert(i);
+    }
+
+    // Initialise splits record
+    splits_.clear();
+    splits_.resize(1, std::unordered_set<bst_uint>());
+  }
+
+  SplitEvaluator* GetHostClone() const override {
+    if (params_.interaction_constraints == "NULL") {
+      // No interaction constraints specified, just return a clone of inner
+      return inner_->GetHostClone();
+    } else {
+      auto c = new InteractionConstraint(
+        std::unique_ptr<SplitEvaluator>(inner_->GetHostClone()));
+      c->params_ = this->params_;
+      c->Reset();
+      return c;
+    }
+  }
+
+  bst_float ComputeSplitScore(bst_uint nodeid,
+                              bst_uint featureid,
+                              const GradStats& left_stats,
+                              const GradStats& right_stats,
+                              bst_float left_weight,
+                              bst_float right_weight) const override {
+    // Return negative infinity score if feature is not permitted by interaction constraints
+    bst_float infinity = std::numeric_limits<bst_float>::infinity();
+    if (!CheckInteractionConstraint(featureid, nodeid)) {
+      return -infinity;
```

Comment:
There's no need to create a variable for that.

Reply:
Fixed